### PR TITLE
Add PHPStan CI step

### PR DESCRIPTION
**Pull request by nickygerritsen**
This adds a PHPStan CI step as discussed in Slack.

It also changes the code to make sure we are PHPStan level 6 compatible, the level suggested by the PHPStan developers for an application.

There are some rules that require us to specify a type for every array parameter, i.e. use `int[]` or `array<int>` instead of `array`, but many of our arrays take `mixed[]` and it didn't make sense to me to add all of these, so I created a baseline that ignores these errors, for now.

### CODE REVIEW

File: phpstan-baseline.neon
Comment (vmcj) [inline phpstan-baseline.neon:1]: I think we normally move such files inside .github/jobs/data.

Comment (nickygerritsen) [reply phpstan-baseline.neon:1]: The advantage of keeping it here is you can locally run
```
lib/vendor/bin/phpstan analyse
```
without specifying where the file is, otherwise you need to point it to that folder. That outweighed the 'normally' for me.

File: webapp/tests/Unit/Controller/API/BaseTestCase.php [hunk 1]
```diff
@@ -60,6 +60,20 @@ protected function setUp(): void
             ->findOneBy(['shortname' => 'demo']);
     }
 
+    /**
+     * Get the contest ID of the demo contest based on the data source setting.
+     *
+     * @return string
+     */
+    protected function getDemoContestId(): string
```
Comment (vmcj) [inline webapp/tests/Unit/Controller/API/BaseTestCase.php:68]: Is this not something which we also want to reuse in non API tests?

Comment (nickygerritsen) [reply webapp/tests/Unit/Controller/API/BaseTestCase.php:68]: If we do we move it (and the property that backs it) to the normal base test case. Now this method was here but the property was in the non-API class, which PHPStan disliked (and so did I).

File: webapp/src/Service/ScoreboardService.php [hunk 2]
```diff
@@ -418,7 +418,7 @@ public function calculateScoreRow(
                 'cid' => $contest->getCid(),
                 'probid' => $problem->getProbid(),
                 'teamSortOrder' => $team->getCategory()->getSortorder(),
-                'submitTime' => $absSubmitTime,
```
Comment (vmcj) [inline webapp/src/Service/ScoreboardService.php:421]: Can you add a TODO so we remember to find which branch is not handled?

File: webapp/tests/Unit/Controller/Jury/JuryControllerTestCase.php [hunk 3]
```diff
@@ -348,13 +350,15 @@ public function testCheckEditEntityAdmin(string $identifier, array $formDataKeys
...
             $crawler = $this->getCurrentCrawler();
+            /** @var \DOMElement $node */
             foreach ($crawler->filter('a') as $node) {
                 if (str_contains($node->nodeValue, $identifier)) {
                     $singlePageLink = $node->getAttribute('href');
                 }
             }
             $this->verifyPageResponse('GET', $singlePageLink, 200);
             $crawler = $this->getCurrentCrawler();
+            /** @var \DOMElement $node */
```
Comment (vmcj) [inline webapp/tests/Unit/Controller/Jury/JuryControllerTestCase.php:361]: You already include `DOMElement` I think

File: webapp/src/Entity/Clarification.php [hunk 4]
```diff
@@ -98,6 +98,9 @@ class Clarification extends BaseApiEntity implements ExternalRelationshipEntityI
     #[Serializer\Exclude]
     private ?Clarification $in_reply_to = null;
 
+    /**
+     * @var Collection<int, Clarification>
```
Comment (vmcj) [inline webapp/src/Entity/Clarification.php:102]: Will we fail CI if/when someone doesn't add such statements?

File: webapp/tests/Unit/Controller/Team/MiscControllerTest.php [hunk 5]
```diff
@@ -148,36 +148,36 @@ public function testChangeContest(bool $withReferrer): void
...
-            self::assertEquals('http://localhost/team',
-                               $this->client->getRequest()->getUri());
-
-            // Go to the scoreboard again.
-            $this->client->request('GET', '/team/scoreboard');
-        }
-
-        // And check that the contest has changed.
-        self::assertSelectorTextContains('.card-header span', 'Test contest for switching');
+//        // Verify we are on the demo contest.
```
Comment (vmcj) [inline webapp/tests/Unit/Controller/Team/MiscControllerTest.php:151]: What is the reason for this?

File: webapp/src/Entity/Clarification.php (hunk 4 above)
Comment (nickygerritsen) [reply webapp/src/Entity/Clarification.php:102]: I believe so. We could disable this rule but it adds a lot of functionality

File: webapp/tests/Unit/Controller/Team/MiscControllerTest.php (hunk 5 above)
Comment (nickygerritsen) [reply webapp/tests/Unit/Controller/Team/MiscControllerTest.php:151]: It’s dead code and phpstan said it is dead code

File: webapp/tests/Unit/Service/ImportExportServiceTest.php [hunk 6]
```diff
@@ -604,9 +604,6 @@ protected function testImportAccounts(int $importCount, ?string $message, bool $
                 if (isset($data['team']['members'])) {
                     self::assertEquals($data['team']['members'], $team->getPublicDescription());
                 }
-                if (isset($data['team']['description'])) {
```
Comment (vmcj) [inline webapp/tests/Unit/Service/ImportExportServiceTest.php:607]: Why do we need to remove this test?

File: webapp/tests/Unit/Controller/Team/MiscControllerTest.php (hunk 5 above)
Comment (vmcj) [reply webapp/tests/Unit/Controller/Team/MiscControllerTest.php:151]: Can you add a TODO to revive the code?

File: webapp/tests/Unit/Service/ImportExportServiceTest.php (hunk 6 above)
Comment (nickygerritsen) [reply webapp/tests/Unit/Service/ImportExportServiceTest.php:607]: It’s never set

File: webapp/tests/Unit/Controller/Team/MiscControllerTest.php (hunk 5 above)
Comment (nickygerritsen) [reply webapp/tests/Unit/Controller/Team/MiscControllerTest.php:151]: It’s there 2 lines above

**Review by vmcj — APPROVED** [review]
Overall I think it is an improvement.

File: webapp/src/Controller/API/JudgehostController.php [hunk 7]
```diff
@@ -224,7 +224,6 @@ public function updateJudgeHostAction(
             throw new BadRequestHttpException('Argument \'enabled\' is mandatory');
         }
 
-        /** @var Judgehost $judgehost */
```
Comment (meisterT) [inline webapp/src/Controller/API/JudgehostController.php:227]: Why are these removed?

File: webapp/src/Service/EventLogService.php [hunk 8]
```diff
@@ -54,6 +55,17 @@ class EventLogService implements ContainerAwareInterface
     // TODO: Add a way to specify when to use external ID using some (DB)
     // config instead of hardcoding it here. Also relates to
     // AbstractRestController::getIdField.
+    /**
```
Comment (meisterT) [inline webapp/src/Service/EventLogService.php:58]: Stuff like this is a little bit annoying

File: webapp/src/Controller/API/JudgehostController.php (hunk 7 above)
Comment (nickygerritsen) [reply webapp/src/Controller/API/JudgehostController.php:227]: Because both phpstorm and phpstan know the type already. We can keep them but then they should change to add `|null` (I.e. they were wrong)

File: webapp/src/Service/EventLogService.php (hunk 8 above)
Comment (nickygerritsen) [reply webapp/src/Service/EventLogService.php:58]: I could drop this, but then we need a small change in the logic below, since it assumed stuff that wasn’t true

File: webapp/src/Controller/API/JudgehostController.php (hunk 7 above)
Comment (meisterT) [reply webapp/src/Controller/API/JudgehostController.php:227]: Then phpstorm has become smarter, but yeah 👍

**Review by meisterT — APPROVED** [review]

File: webapp/src/Controller/API/JudgehostController.php (hunk 7 above)
Comment (nickygerritsen) [reply webapp/src/Controller/API/JudgehostController.php:227]: It has indeed. AI is taking over the world

File: webapp/src/Service/EventLogService.php (hunk 8 above)
Comment (nickygerritsen) [reply webapp/src/Service/EventLogService.php:58]: I will see what it takes to not do this since it’s indeed a bit verbose

Comment (nickygerritsen) [reply webapp/src/Service/EventLogService.php]: I changed this to another way of doing it, this seems shorter/better/less of a hassle.
